src/seismometer/data/loader/pipeline.py

```diff
@@ -0,0 +1,158 @@
+
+import pandas as pd
+
+from seismometer.configuration import ConfigProvider
+
+logger = logging.getLogger("seismometer")
+
+ConfigOnlyHook: TypeAlias = Callable[[ConfigProvider], pd.DataFrame]
+ConfigOnlyHook.__doc__ = """
+TypeAlias for a callable taking a ConfigProvider, which returns a DataFrame.
+
+An example is loading the predictions from a parquet file, where the ConfigProvider
+provides the path to the file which is loaded as a dataframe.
+"""
+
+ConfigFrameHook: TypeAlias = Callable[[ConfigProvider, pd.DataFrame], pd.DataFrame]
+ConfigFrameHook.__doc__ = """
+TypeAlias for a callable taking a ConfigProvider and a DataFrame, which returns a DataFrame.
+
+An example is ensuring types of a loaded predictions dataframe, where ConfigProvider
+provides some metadata that can then be applied to the dataframe returing the transformed object.
+"""
+
+
+MergeHook: TypeAlias = Callable[[ConfigProvider, pd.DataFrame, pd.DataFrame], pd.DataFrame]
+MergeHook.__doc__ = """
+TypeAlias for a callable taking a ConfigProvider and two DataFrames, which returns a DataFrame.
+
+An example is uses ConfigProvider to understand the relevant keys, then
+merging events from the first dataframe argument onto the last argument (predictions)
+returning the combined dataframe.
+"""
+
+
+def _passthru_framehook(config: ConfigProvider, dataframe: pd.DataFrame) -> pd.DataFrame:
+    """A generic pass-through used for hooks that don't need to modify the dataframe."""
+    return dataframe
+
+
+def _passthru_mergehook(config: ConfigProvider, eventframe: pd.DataFrame, dataframe: pd.DataFrame) -> pd.DataFrame:
+    """A generic pass-through used for hooks that don't need to modify the dataframe."""
+    return dataframe
+
+
+class SeismogramLoader:
+    """
+    A data loading pipeline using three types of hooks:
+
+    * load predictions [ConfigOnlyHook]
+    * transform (type) the loaded predictions [ConfigFrameHook]
+    * load events [ConfigOnlyHook]
+    * transform (type) the loaded events [ConfigFrameHook]
+    * merge events onto predictions [MergeHook]
+
+    Each step is expected to return a dataframe, chaining the steps to get the frame driving a loaded Seismogram.
+    """
+
+    def __init__(
+        self,
+        config: ConfigProvider,
+        prediction_fn: Optional[ConfigOnlyHook] = None,
+        event_fn: Optional[ConfigOnlyHook] = None,
+        post_predict_fn: Optional[ConfigFrameHook] = None,
+        post_event_fn: Optional[ConfigFrameHook] = None,
+        merge_fn: Optional[MergeHook] = None,
+    ):
+        """
+        Initialize a data loading pipeline of functions returning a dataframe for a Seismogram session.
+
+        Parameters
+        ----------
+        config : ConfigProvider
+            The loaded configuration object.
+        prediction_fn : ConfigOnlyHook
+            A callable taking a ConfigProvider and returning a dataframe.
+            Used to load a (predictions) dataframe based on configuration.
```

Comment:
Can we add comment about this being ignored if load_data is called with a dataframe?

Reply:
Ammended here and event_fn